### ISSUE TITLE
newly imported channels don't show up in Learn without restart

### ISSUE THREAD

**Issue by indirectlylit** [learningequality/kolibri#6653]
### Observed behavior

Observed on a few occasions:

* Import a channel, and let it complete
* Navigate to Learn
* The channel is not visible

### Expected behavior

Channels visibility should be consistent between Device and Learn

### User-facing consequences

Confusion and frustration. I tried reloading many times to no avail.

Eventually restarted the server and then the channel was visible.

### Errors and logs

None visible

### Steps to reproduce

* Import channel
* navigate to Learn

### Context

* 0.13.1
* dev server
* chrome

**Comment by jonboiser** [comment]
@rtibbles I've noticed this when you "import more", because of the `channel_stats` caching (the only thing we haven't done in #6533 and #6626). For a brand new channel, it may be because of caching on the `/api/contentnode` endpoint?

**Comment by rtibbles** [comment]
The channel stats doesn't get used for visibility on learn though - it seems more likely that this is an issue with the etag generation or not being updated.

What browser is it this happening in?

**Comment by rtibbles** [comment]
Another possibility is that there has been a regression due to the updates, which is causing the content cache key not to get regenerated. Or it is being regenerated somewhere but the cache not being updated.

**Comment by rtibbles** [comment]
Chain of events that should be happening:

New content gets imported
ContentCacheKey gets updated
Cached value of ContentCacheKey also gets updated
etag for all HTML pages gets recalculated
HTML page reloads with new ContentCacheKey value bootstrapped in
Content API calls have this new ContentCacheKey as a GET PARAM, so hit the server rather than browser  cache, and update

**Comment by benjaoming** [comment]
I read through the implemented behavior and didn't spot anything, looking at HTTP response headers now.

The response of `/learn/` correctly states max-age=0, which means:

> max-age=0 indicates that a cache entry is stale and requires re-validation, but does not prevent caching. Often browsers only validate resources once per browser session, so the content may not get updated until the site is visited in a new session.

This means that a re-validation with `If-Not-Modified` is desired.

Observation: The expiration date of endpoints such as `/api/content/channel/` and `/api/content/contentnode_slim/` is set to a year in the future.  I'm not sure if this is intended or indeed if it means anything for an `application/json` request.

Example:

```
HTTP/1.1 200 OK
Content-Type: application/json
Vary: Accept, Cookie
Allow: GET, POST, HEAD, OPTIONS
Expires: Sat, 20 Mar 2021 21:53:10 GMT
Cache-Control: max-age=31556926
X-Frame-Options: SAMEORIGIN
Content-Length: 1280
Date: Fri, 20 Mar 2020 16:04:24 GMT
Server: 0.0.0.0
```

**Comment by benjaoming** [comment]
I can see that API requests do not provide `If-Not-Modified`, so I assume that's why we use a URL parameter -- I'm not sure if there is a case where the previous content cache key prevails as a URL parameter for the API views?

![image](https://user-images.githubusercontent.com/374612/77184250-d60a5f80-6acf-11ea-80be-652eff741f03.png)

**Comment by rtibbles** [comment]
> Observation: The expiration date of endpoints such as /api/content/channel/ and /api/content/contentnode_slim/ is set to a year in the future. I'm not sure if this is intended or indeed if it means anything for an application/json request.

This is intentional, to provide long term in browser caching of these resources. The ContentCacheKey is meant to be invalidating this cache when things change.

**Comment by rtibbles** [comment]
> I can see that API requests do not provide If-Not-Modified, so I assume that's why we use a URL parameter -- I'm not sure if there is a case where the previous content cache key prevails as a URL parameter for the API views?

The content cache key is set in Javascript - drawing from the value that is bootstrapped into the HTML page served by Django. Is the content cache key GET param updating? If not, then that is the cause of the issue.

**Comment by benjaoming** [comment]
I cannot reproduce the issue, so I'm approaching it purely theoretical -- I'm guessing we could close this issue if we could fix one or more theoretical gaps in how the content cache key is updated and transmitted.

**Comment by rtibbles** [comment]
I am wondering if it might be a specific browser caching issue, as both @jonboiser and @indirectlylit have reported it.

**Comment by benjaoming** [comment]
Any chance that this behavior is confined to kolibri-server? Am wondering this in case different instances of the Django cache backend results in the wrong `contentCacheKey` being handed out.

**Comment by indirectlylit** [comment]
no this is reproducible locally with the dev server - see the context in the description

**Comment by benjaoming** [comment]
Sorry @indirectlylit - I had already deleted the comment, was a glitch in a thought train.

I'm adding the following notes:

* I still cannot reproduce this. Tried a dev server with Firefox+Chromium on Ubuntu and Google Chrome on Mac.
* #6682 changes the cache key and the cached cache key to both be integers. Before, they were int and float, and that does not match and creates a theoretical 1 second overlap that I can't figure out a theoretical significance for. So I recommend scratching it as a potential fix.

**Comment by indirectlylit** [comment]
This is an issue with how caches work in different contexts:

* dev server uses separate processes and non-shared memory so reproduces bug
* production server uses threads and shared memory so bug wouldn't reproduce
* cloud kolibri and kolibri server use redis which works cross-process

removing from milestone and reducing severity, but we should still fix eventually

recommended fix: use `CrossProcessCache` for `cache_key` caching (whooo!)

Workaround in the meantime: trigger a server restart by making a trivial change to a python file (cc @jonboiser)

**Comment by jonboiser** [comment]
Should be fixed by #7766 🎉

**Comment by indirectlylit** [comment]
hooray!
